### PR TITLE
[web][api][SIMS #1124]Reorganize student API (Part 3)

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- All the endpoints under the controller `student.controller.ts` were moved/removed;
- Created the common pagination parameters to be used whatever we need a paginated API endpoint `PaginationOptionsAPIInDTO` and created the generic paginated result DTO `PaginatedResultsAPIOutDTO`, as shown below. The class `ApplicationPaginationOptionsAPIInDTO` is extended from `PaginationOptionsAPIInDTO` to allow complete validation of all query parameters including the possible sort fields.
image
- Moved GET `education-program` from student controller to the `education-program` student exclusive controller `education-program.students.controller.ts`;
- API endpoint `@Get("restriction")` to be removed once the PR #1233 is merged.

### CODE REVIEW

**Review by henryoforeh-aot — APPROVED** [review]

File: sources/packages/api/src/route-controllers/student/student.aest.controller.ts [hunk 1]
```diff
@@ -197,4 +226,25 @@ export class StudentAESTController extends BaseController {
...
+
+  /**
+   * Get the list of application that belongs to a student on a summary view format.
+   * @param pagination options to execute the pagination.
+   * @returns student application list with total count.
+   */
+  @Get(":studentId/application-summary")
+  @ApiNotFoundResponse({ description: "Student does not exists." })
+  async getStudentApplicationSummary(
+    @Param("studentId") studentId: number,
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/student/student.aest.controller.ts:238]: pls add it in the comment

File: sources/packages/api/src/route-controllers/student/student.controller.service.ts [hunk 2]
```diff
@@ -122,4 +134,49 @@ export class StudentControllerService {
...
   }
+
+  /**
+   * API to fetch all the applications that belong to student.
+   * This API will be used by students.
+   * @param pagination options to execute the pagination.
+   * @returns student application list with total count.
+   */
+  async getStudentApplicationSummary(
+    studentId: number,
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.service.ts:145]: pls add this in the comment

File: sources/packages/api/src/route-controllers/student/student.controller.service.ts [hunk 3]
```diff
@@ -122,4 +134,49 @@ export class StudentControllerService {
...
+    pagination: ApplicationPaginationOptionsAPIInDTO,
+  ): Promise<PaginatedResultsAPIOutDTO<ApplicationSummaryAPIOutDTO>> {
+    const [applications, count] =
+      await this.applicationService.getAllStudentApplications(
+        studentId,
+        pagination,
+      );
+
+    return {
+      results: applications.map((application: Application) => {
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.service.ts:155]: i think `results: applications.map((application: Application) => 
        this.transformToApplicationSummaryDTO(application);
      ),` shoudl also work

File: sources/packages/api/src/services/student/student.service.ts [hunk 4]
```diff
@@ -303,17 +303,16 @@ export class StudentService extends RecordDataModelService<Student> {
...
   /**
-   * Search the student based on the search criteria.
-   * @param firstName firsName of the student.
-   * @param lastName lastName of the student.
-   * @param appNumber application number of the student.
-   * @returns Searched student details.
+   * Search students based on the search criteria.
+   * @param searchCriteria options to search by firstName,
+   * lastName or appNumber.
+   * @returns searched student details.
```
Comment (ann-aot) [inline sources/packages/api/src/services/student/student.service.ts:309]: is it student details or list of students

File: sources/packages/api/src/services/student/student.service.ts [hunk 5]
```diff
@@ -329,22 +328,22 @@ export class StudentService extends RecordDataModelService<Student> {
...
       searchQuery.andWhere("user.lastName Ilike :lastName", {
-        lastName: `%${lastName}%`,
+        lastName: `%${searchCriteria.lastName}%`,
       });
     }
-    if (appNumber) {
+    if (searchCriteria.appNumber) {
       searchQuery
         .andWhere("application.applicationNumber Ilike :appNumber")
         .andWhere("application.applicationStatus != :overwrittenStatus")
```
Comment (ann-aot) [inline sources/packages/api/src/services/student/student.service.ts:344]: I know, its not part of the PR. shouldn't `.andWhere("application.applicationStatus != :overwrittenStatus")` be outside the `if`

File: sources/packages/web/src/services/http/StudentApi.ts [hunk 6]
```diff
@@ -111,37 +88,31 @@ export class StudentApi extends HttpBaseClient {
...
-   * API client for student search.
-   * @param appNumber
-   * @param firstName
-   * @param lastName
-   * @returns
+   * Search students based on the search criteria.
+   * @param appNumber application number.
+   * @param firstName students first name (given names).
+   * @param lastName student last name
+   * @returns searched student details.
```
Comment (ann-aot) [inline sources/packages/web/src/services/http/StudentApi.ts:95]: is its a list of students or details of the student

**Review by ann-aot — COMMENTED** [review]
added some comments

File: sources/packages/api/src/route-controllers/models/pagination.dto.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  /**
+   * Criteria to be used to filter the records.
+   */
+  @IsOptional()
+  searchCriteria?: string;
+}
+
+export class ApplicationPaginationOptionsAPIInDTO extends PaginationOptionsAPIInDTO {
+  @IsIn(["status", "applicationNumber"])
+  sortField?: string;
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/models/pagination.dto.ts:38]: what about IsOptional() ?

Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/models/pagination.dto.ts:38]: Or else we can override it to be mandatory if it is mandatory for the implementing DTO.

image

File: sources/packages/api/src/services/student/student.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/student/student.service.ts:309]: It is a subset with firstName, lastName, birthDate. What is the point here?

File: sources/packages/api/src/services/student/student.service.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/student/student.service.ts:344]: It seems that if the application will be searched then look only into the not overwrittenStatus. Either way, it seems weird t me too.
@guru-aot can you refresh our memory here 😉

File: sources/packages/web/src/services/http/StudentApi.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/services/http/StudentApi.ts:95]: It is a subset with firstName, lastName, birthDate. What is the point here?

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts [hunk 8]
```diff
@@ -128,3 +133,16 @@ export class AESTFileUploadToStudentAPIInDTO {
   @ArrayMinSize(1)
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/models/student.dto.ts:133]: We can remove this class-validator in out DTO.

File: sources/packages/api/src/route-controllers/models/pagination.dto.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/models/pagination.dto.ts:38]: Yes, it would be good to have it added.

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts [hunk 9]
```diff
@@ -128,3 +133,16 @@ export class AESTFileUploadToStudentAPIInDTO {
...
+ * DTO object application summary info.
+ */
+export class ApplicationSummaryAPIOutDTO {
+  applicationNumber: string;
+  studyStartPeriod: string;
+  studyEndPeriod: string;
+  id: number;
+  applicationName: string;
+  submitted?: Date;
+  status: string;
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/models/student.dto.ts:147]: It is the application status? We can use enum here if so.

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts (hunk 8 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/models/student.dto.ts:133]: Sorry, I am not following, it is an In DTO? Am I missing something?
image

File: sources/packages/api/src/services/student/student.service.ts (hunk 4 above)
Comment (ann-aot) [reply sources/packages/api/src/services/student/student.service.ts:309]: mm. I see that promise is a student array, so it should be a list of student's detail right?

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts [hunk 10]
```diff
@@ -17,41 +20,43 @@ export class ContactInformationAPIOutDTO {
...
- * or updating the student profile.
- * Validations not added to DTO because
- * they are going to be handled by the
- * Form.IO dryRun validation.
+ * Updates the student information that the student is allowed to change
+ * in the solution. Other data must be edited externally (e.g. BCSC).
  */
-export interface SaveStudentDto extends AddressDetailsAPIInDTO {
+export class UpdateStudentAPIInDTO extends AddressDetailsAPIInDTO {
+  @Length(10, 20)
```
Comment (guru-aot) [inline sources/packages/api/src/route-controllers/student/models/student.dto.ts:40]: 👍

File: sources/packages/api/src/route-controllers/student/student.aest.controller.ts [hunk 11]
```diff
@@ -179,6 +189,25 @@ export class StudentAESTController extends BaseController {
...
+  async searchStudents(
+    @Query() searchCriteria: AESTStudentSearchAPIInDTO,
+  ): Promise<SearchStudentAPIOutDTO[]> {
+    const searchStudentApplications =
+      await this.studentService.searchStudentApplication(searchCriteria);
+    return searchStudentApplications.map((eachStudent: Student) => ({
+      id: eachStudent.id,
+      firstName: eachStudent.user.firstName,
+      lastName: eachStudent.user.lastName,
+      birthDate: getISODateOnlyString(eachStudent.birthDate),
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/student.aest.controller.ts:207]: As per figma, all the dates at UI must be in this format I believe. From here we are sending the ISO format. Are we expecting the client side to transform it again?
image

File: sources/packages/api/src/route-controllers/student/student.aest.controller.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.aest.controller.ts:239]: Added

File: sources/packages/api/src/route-controllers/student/student.controller.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.service.ts:146]: Added

File: sources/packages/api/src/route-controllers/student/student.controller.service.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.service.ts]: Changed

File: sources/packages/api/src/services/student/student.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/student/student.service.ts]: Changed

File: sources/packages/api/src/route-controllers/models/pagination.dto.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/models/pagination.dto.ts:39]: I added it for now to keep it in sync with the service model also (`PaginationOptions`).

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts (hunk 9 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/models/student.dto.ts]: Changed

File: sources/packages/api/src/route-controllers/student/student.aest.controller.ts (hunk 11 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.aest.controller.ts:207]: As we talked about in the past it would be best to API return the date itself (and as a string) when we have the date-only.
And yes, the Vue is supposed to do the transformation, it was not, I forgot, and I just added 😄 
image

Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.aest.controller.ts:207]: The comma is now removed from the dates formatted with `dateOnlyLongString`.

**Review by ann-aot — APPROVED** [review]
Thanks for doing the changes. 👍

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/student/models/student.dto.ts:133]: My bad this looked like it was added to OutDTO😊

image

File: sources/packages/api/src/route-controllers/student/student.aest.controller.ts (hunk 11 above)
Comment (dheepak-aot) [reply sources/packages/api/src/route-controllers/student/student.aest.controller.ts:207]: thanks for doing that @andrewsignori-aot 👍

File: sources/packages/api/src/route-controllers/student/student.aest.controller.ts [hunk 12]
```diff
@@ -197,4 +226,26 @@ export class StudentAESTController extends BaseController {
     }
     return student;
   }
+
+  /**
+   * Get the list of application that belongs to a student on a summary view format.
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/student.aest.controller.ts:231]: list of application**s**?

File: sources/packages/api/src/services/student/student.service.ts (hunk 5 above)
Comment (guru-aot) [reply sources/packages/api/src/services/student/student.service.ts:344]: This is a student search and, if the user searches with just application number, then we search for the application which are all not in the overwritten status. If we have it outside the if, then the search happens with respect to user table , so the application status is irrelevant.

File: sources/packages/api/src/route-controllers/student/student.controller.service.ts [hunk 13]
```diff
@@ -122,4 +134,50 @@ export class StudentControllerService {
       validSin: student.sinValidation.isValidSIN,
     };
   }
+
+  /**
+   * API to fetch all the applications that belong to student.
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.service.ts:139]: The comment says API. It may be removed as controller service is not an API by itself unless used by a controller.

Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.service.ts:139]: It was copied as it was and I particularly don't like these comments also because they are not accurate. I will change it.

File: sources/packages/api/src/route-controllers/student/student.controller.service.ts [hunk 14]
```diff
@@ -122,4 +134,50 @@ export class StudentControllerService {
...
+    return {
+      results: applications.map((application: Application) =>
+        this.transformToApplicationSummaryDTO(application),
+      ),
+      count,
+    };
+  }
+
+  /**
+   * Util to transform application entity model to ApplicationSummaryDTO.
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.service.ts:164]: DTO name can be removed from the comment(ApplicationSummaryDTO)

Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.service.ts]: Done

File: sources/packages/api/src/route-controllers/student/student.controller.service.ts (hunk 13 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.service.ts]: Done

File: sources/packages/api/src/route-controllers/student/student.aest.controller.ts (hunk 12 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.aest.controller.ts]: Done

**Review by guru-aot — APPROVED** [review]
LGTM, Nice work @andrewsignori-aot

File: sources/packages/api/src/route-controllers/student/student.students.controller.ts [hunk 15]
```diff
@@ -60,19 +81,110 @@ export class StudentStudentsController extends BaseController {
...
+  @Post()
+  @ApiUnprocessableEntityResponse({
+    description:
+      "There is already a student associated with the user or the request is invalid.",
+  })
+  @RequiresStudentAccount(false)
+  async create(
+    @UserToken() studentUserToken: StudentUserToken,
+    @Body() payload: CreateStudentAPIInDTO,
+  ): Promise<void> {
```
Comment (dheepak-aot) [inline sources/packages/api/src/route-controllers/student/student.students.controller.ts:107]: For Post APIs we are returning the primary identifier of the created resource.

File: sources/packages/web/src/services/StudentService.ts [hunk 16]
```diff
@@ -1,24 +1,16 @@
...
   AESTFileUploadToStudentAPIInDTO,
   AESTStudentFileAPIOutDTO,
+  CreateStudentAPIInDTO,
+  SearchStudentAPIOutDTO,
   StudentFileUploaderAPIInDTO,
   StudentUploadFileAPIOutDTO,
-} from "./http/dto/Student.dto";
+  UpdateStudentAPIInDTO,
+  StudentRestrictionAPIOutDTO,
+} from "./http/dto";
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/StudentService.ts:13]: @/services/http/dto ?

File: sources/packages/web/src/services/ApplicationService.ts [hunk 17]
```diff
@@ -9,11 +9,14 @@ import {
...
   DEFAULT_PAGE_NUMBER,
-  StudentApplicationAndCount,
   ApplicationIdentifiersDTO,
 } from "@/types";
 import { MORE_THAN_ONE_APPLICATION_DRAFT_ERROR } from "@/types/contracts/ApiProcessError";
 import ApiClient from "../services/http/ApiClient";
+import {
+  ApplicationSummaryAPIOutDTO,
+  PaginatedResultsAPIOutDTO,
+} from "./http/dto";
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/ApplicationService.ts:19]: @ import

File: sources/packages/web/src/services/http/dto/Student.dto.ts [hunk 18]
```diff
@@ -54,3 +74,38 @@ export interface AESTStudentFileAPIOutDTO extends StudentUploadFileAPIOutDTO {
...
+  applicationName: string;
+  submitted?: Date;
+  status: string;
+}
+
+/**
+ * Student restriction DTO.
+ * This object is returned by controller.
+ */
+export interface StudentRestrictionAPIOutDTO {
```
Comment (dheepak-aot) [inline sources/packages/web/src/services/http/dto/Student.dto.ts:102]: Thanks for doing this. I know you have merged @ann-aot PR to do this.

File: sources/packages/web/src/components/aest/StudentApplications.vue [hunk 19]
```diff
@@ -184,27 +192,14 @@ export default {
...
-              props.studentId,
-              page,
-              pageCount,
-              sortField,
-              sortOrder,
-            );
-          break;
-      }
+      applicationsAndCount.value =
+        await ApplicationService.shared.getStudentApplicationSummary(
```
Comment (dheepak-aot) [inline sources/packages/web/src/components/aest/StudentApplications.vue:196]: This service method can use the PaginationOptions interface instead of 5 arguments.

File: sources/packages/api/src/route-controllers/student/student.students.controller.ts (hunk 15 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/student/student.students.controller.ts:107]: I missed that while moving the code. I will add it.

File: sources/packages/web/src/components/aest/StudentApplications.vue (hunk 19 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/components/aest/StudentApplications.vue:196]: It is the Web project and I will not refactor it as part of this PR.

**Review by dheepak-aot — APPROVED** [review]
Thanks for doing the changes @andrewsignori-aot 👍
